scalableminds/webknossos
# Fix rendering missing data black and wrong segment id being displayed

This PR tackles two issues:
1. When "Render missing data black" was turned off and the data of the fallback mag for the segmentation layer was visible, all segments had id 0. This was due to not looking up the segment id in the fallback bucket. The data was not looked up, because all buckets from a non-existing magnification for the data layer currently in question (e.g. segmentation layer) were null buckets. The previous code assumed, that if we have null buckets anyway, looking up in other mags does not benefit at all. But this is wrong in case the above-mentioned option is turned off. Thus this pr simply changes this assumption, if the "Render missing data black"-option is turned off.
2. When the "Render missing data black"-option was turned on and a layer does not have all mags, the existing mags were only rendered as fallback data if the current mag was only one zoomStep lower. Although it should be expected that the fallback data should be rendered up to the mag that is three steps lower. This PR removes this restriction. The restriction was previously added, because, with each step, the fallback and the rendered mag are apart, the number of entries in the lookup table grows exponentially, as each bucket in the actually rendered mag needs an entry added to the lookup table. But this overall does not matter as when there was actually data available for the currently rendered magnification, each bucket in the current mag would also need an entry in the lookup table. Thus in both cases the number of entries in the lookup table does not change. 
In the first case, there are simply much fewer buckets of the fallback mag (exponentially decreasing with the mag difference) that are added for many buckets in the rendered mag (exponentially increasing with the mag difference) to the lookup table. Thus the exponential increase and decrease should cancel out each other.

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
Better test locally. First, read the whole steps. It would be good if you have another dataset that has a segmentation for all 16 mags and a color layer with all 16 mags. This might show some edge cases that did not appear with the ROI dataset.
- First set up a new ROI Dataset for testing. For that dataset create a copy of the color layer and only keep mag 16. Then also delete all mags of the segmentation layer except for mag 16.
- Go to the dataset list, open the settings of the ROI dataset and save the suggested setting.
- View the dataset and turn off the original color layer and the segmentation layer in the layer settings. Then turn off the "Render missing data black" setting.
- You shouldn't be able to see data in mag 1, 32. But you should be able to see data in the other mags.
- The same should be the case for the manipulated color layer. To test this, turn on the color layer and turn of the segmentation layer.
- Only make the unchanged color layer visible. Here you should see data in each resolution.
- Now turn on the unchanged color layer and the segmentation layer. Check whether you get a segmentation id for any hovered segment in each mag, in which the segmentation is visible. If the magnifications where the segmentation is visible, the displayed segment id in the status bar should be 0. 

### Issues:
- fixes #5641 #5656

------
